### ISSUE TITLE
[RFC] Add `availabilityPredicate` to the promotion

### ISSUE THREAD

**Issue by IKarbowiak** [saleor/saleor#12925]
## Problem

Allow customizing where the promotion is active. Introduce the `availabilityPredicate` that will contain the condition that must be fulfilled to consider the promotion as active.

Blocked by: https://github.com/saleor/saleor/issues/12539

---

## 1. General Assumptions

- Allow defining `availabilityPredicate` on `Promotion`
- Availability predicate should allow defining on which weekday and thick what frequency the promotion is active
- Update the promotion-based calculations
- Ensure the discounted prices are recalculated after changing the availability option
- Update the celery task for discounted price recalculations to include the promotion availability changes 

### Examples
- Available every Wednesday
```josn
{
 "promotion": {
  "availabilityPredicate": {
  	"weekDay": ["WEDNESDAY"]
  },
 }
}
```

- Available on Sunday, and Saturday every 4 weeks and on Friday every 2 weeks
```json
{
 "promotion": {
  "availabilityPredicate": {
     "AND": [
      {
  	  "weekDay": ["SATURDAY", "SUNDAY"],
  	  "weeksFrequency": 4
      },
      {
  	  "weekDay": ["FRIDAY"],
  	  "weeksFrequency": 2
      }
     ]
  },
 }
}
```

### Acceptance Criteria

- [ ] The products should show correct prices depend on promotion availability

### Mutations
- Allow defining the `availabilityPredicate` in the promotion mutations: `PromotionCreate `, `PromotionUpdate`
  - Call the discounted price recalculation when the availability has changed
```graphql
input PromotionCreateInput @doc(category: "Discounts") {
  ...
  availabilityPredicate: AvailabilityPredicateInput
}

input PromotionUpdateInput @doc(category: "Discounts") {
  ...
  availabilityPredicate: AvailabilityPredicateInput
}

type AvailabilityPredicateInput @doc(category: "Discounts"){
  weekDay: [WeekDayEnum!]
  weeksFrequency: Int
  AND: [AvailabilityPredicateInput]
  OR: [AvailabilityPredicateInput]
}
```

### Types
- Extend `Promotion` with `availabilityPredicate`
```graphql
type Promotion implements Node & ObjectWithMetadata @doc(category: "Discounts") {
  ...
  availabilityPredicate: AvailabilityPredicate
}

type AvailabilityPredicate @doc(category: "Discounts"){
  weekDay: [WeekDayEnum!]
  weeksFrequency: Int
  AND: [AvailabilityPredicate]
  OR: [AvailabilityPredicate]
}

enum WeekDayEnum @doc(category: "Discounts") {
  MONDAY
  THURSDAY
  WEDNESDAY
  THURSDAY
  FRIDAY
  SUNDAY
  SATURDAY
}
```

## 3. Database changes

- Extend `Promotion` model with `availabilityPredicate`
```python
class Promotion:
    availability_predicate: JSONField
```

## 4. To Do list

- [ ] Add `availabilityPredicate` type
- [ ] Allow defining `availabilityPredicate` in promotion mutations
- [ ] Extend the `Promotion` model
- [ ] Update the recalculations
- [ ] Update the scheduled task for discounted price recalculations

## 5. To Test

- [ ]
- [ ]

**Comment by IKarbowiak** [comment]
Postponed for now.
